bug: Non-Deterministic Test HotReloadIT.shouldModifyPortAddressedVcWithSamePort

A test failed non-deterministically in GitHub Actions:

## failed test

`HotReloadIT.shouldModifyPortAddressedVcWithSamePort`

[failure](https://github.com/kroxylicious/kroxylicious/actions/runs/31041068769/job/92427006230?pr=4587)

## stack trace

```
 [INFO] Running io.kroxylicious.it.HotReloadIT
2026-08-05 19:58:26 WARN  <Thread-288> i.k.p.i.n.EndpointRegistry - Registration error 
java.util.concurrent.CompletionException: java.lang.RuntimeException: bind to <any>:41456 failed
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:994)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:72)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: bind to <any>:41456 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	... 3 more
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:26 WARN  <main> i.k.p.i.r.AddCluster - reconfigure: gateway registration failed; rolling back 
java.lang.RuntimeException: bind to <any>:41456 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:26 WARN  <main> i.k.p.StartupShutdownLogger - Virtual cluster reached terminal stopped state due to failure, proxy shutdown required 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:30 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:31 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:32 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:33 WARN  <multiThreadIoEventLoopGroup-1433-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:34 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:34 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:34 WARN  <multiThreadIoEventLoopGroup-1433-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:35 WARN  <multiThreadIoEventLoopGroup-1433-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:45 WARN  <Thread-298> i.k.p.i.n.EndpointRegistry - Registration error 
java.util.concurrent.CompletionException: java.lang.RuntimeException: bind to <any>:46597 failed
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:994)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:72)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: bind to <any>:46597 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	... 3 more
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:45 WARN  <main> i.k.p.i.r.AddCluster - reconfigure: gateway registration failed; rolling back 
java.lang.RuntimeException: bind to <any>:46597 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:45 WARN  <main> i.k.p.StartupShutdownLogger - Virtual cluster reached terminal stopped state due to failure, proxy shutdown required 
2026-08-05 19:58:51 WARN  <Thread-304> i.k.p.i.n.EndpointRegistry - Registration error 
java.util.concurrent.CompletionException: java.lang.RuntimeException: bind to <any>:42321 failed
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:994)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:72)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: bind to <any>:42321 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	... 3 more
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:51 WARN  <main> i.k.p.i.r.AddCluster - reconfigure: gateway registration failed; rolling back 
java.lang.RuntimeException: bind to <any>:42321 failed
	at io.kroxylicious.proxy.internal.net.NetworkBindRequest.lambda$performBindingOperation$0(NetworkBindRequest.java:73)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed with error(-98): Address already in use
2026-08-05 19:58:51 WARN  <main> i.k.p.StartupShutdownLogger - Virtual cluster reached terminal stopped state due to failure, proxy shutdown required 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:54 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:55 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:56 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:57 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-2> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:58 WARN  <multiThreadIoEventLoopGroup-1465-3> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-4> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
2026-08-05 19:58:59 WARN  <multiThreadIoEventLoopGroup-1465-1> i.k.p.i.KafkaProxyInitializer - Exception resolving Virtual Cluster Binding 
[ERROR] Tests run: 13, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 42.67 s <<< FAILURE! -- in io.kroxylicious.it.HotReloadIT
[ERROR] io.kroxylicious.it.HotReloadIT.shouldModifyPortAddressedVcWithSamePort(KafkaCluster) -- Time elapsed: 1.506 s <<< FAILURE!
org.assertj.core.error.AssertJMultipleFailuresError: 

Multiple Failures (1 failure)
-- failure 1 --
[ReconfigureResult should have no errors for a clean same-port modify] 
Expecting value to be false but was true
at HotReloadIT.lambda$shouldModifyPortAddressedVcWithSamePort$10(HotReloadIT.java:459)
	at io.kroxylicious.it.HotReloadIT.shouldModifyPortAddressedVcWithSamePort(HotReloadIT.java:457)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


```

--- a/kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java
+++ b/kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java
@@ -65,7 +65,8 @@ class HotReloadIT extends BaseIT {
     private static final Logger LOGGER = LoggerFactory.getLogger(HotReloadIT.class);
 
     private static final int PORT_REUSE_BOOTSTRAP = 9292;
-    private static final String PORT_REUSE_LOCK = "localhost:" + PORT_REUSE_BOOTSTRAP;
+    private static final int PORT_RELOCATE_TARGET_BOOTSTRAP = PORT_REUSE_BOOTSTRAP + 10;
+    private static final int PORT_SAME_MODIFY_BOOTSTRAP = 9291;
 
     private static final String VC_BASELINE_NAME = "vc-baseline";
     private static final String VC_OUTGOING_NAME = "vc-outgoing";
@@ -344,7 +345,7 @@ class HotReloadIT extends BaseIT {
     }
 
     @Test
-    @ResourceLock(PORT_REUSE_LOCK)
+    @ResourceLock("localhost:" + PORT_REUSE_BOOTSTRAP)
     void shouldSupportPortReuseAcrossReconfigures(@BrokerCluster KafkaCluster cluster) throws Exception {
         var startingConfig = portConfig(
                 portVc(cluster, "vc-retain"),
@@ -430,15 +431,13 @@ class HotReloadIT extends BaseIT {
     }
 
     @Test
+    @ResourceLock("localhost:" + PORT_SAME_MODIFY_BOOTSTRAP)
     void shouldModifyPortAddressedVcWithSamePort(@BrokerCluster KafkaCluster cluster) throws Exception {
         // Same-port modify is the tight-timing case: ReplaceCluster's internal unbind happens
         // immediately before its rebind of the SAME port. Triggered here by flipping logNetwork
         // (a runtime-observable field whose change `VirtualCluster.sameAs` reports as a modify
         // but which doesn't affect client behaviour, so the cluster keeps working).
-        //
-        // Starts on port 0 so the OS assigns a port safely; after startup we capture the
-        // bound port and reconfigure with that explicit port to guarantee same-port rebind.
-        var startingConfig = portConfig(portVc(cluster, "vc-modify"));
+        var startingConfig = portConfig(portVc(cluster, "vc-modify", PORT_SAME_MODIFY_BOOTSTRAP));
 
         var testerBuilder = KroxyliciousConfigUtils.baseConfigurationBuilder()
                 .addToVirtualClusters(startingConfig.virtualClusters().toArray(new VirtualCluster[0]));
@@ -447,11 +446,9 @@ class HotReloadIT extends BaseIT {
             // Given
             String topic = tester.createTopic("vc-modify");
             assertProduceConsumeRoundTrip(tester, "vc-modify", topic, "given-pre-modify");
-            int port = boundPort(tester, "vc-modify");
 
             // When
-            var afterConfig = portConfig(portVcWithLogNetwork(cluster, "vc-modify", port, true));
-            LOGGER.info("Reconfiguring to modify vc-modify (same port {}, logNetwork=true)", port);
+            var afterConfig = portConfig(portVcWithLogNetwork(cluster, "vc-modify", PORT_SAME_MODIFY_BOOTSTRAP, true));
             assertThat(tester.reconfigure(afterConfig))
                     .succeedsWithin(RECONFIGURE_TIMEOUT)
                     .satisfies(rr -> assertThat(rr.hasErrors())
@@ -464,10 +461,10 @@ class HotReloadIT extends BaseIT {
     }
 
     @Test
-    @ResourceLock(PORT_REUSE_LOCK)
+    @ResourceLock("localhost:" + PORT_REUSE_BOOTSTRAP)
+    @ResourceLock("localhost:" + PORT_RELOCATE_TARGET_BOOTSTRAP)
     void shouldModifyPortAddressedVcWithDifferentPort(@BrokerCluster KafkaCluster cluster) throws Exception {
         int relocateFromPort = PORT_REUSE_BOOTSTRAP;
-        int relocateToPort = PORT_REUSE_BOOTSTRAP + 10;
         var testerBuilder = KroxyliciousConfigUtils.baseConfigurationBuilder()
                 .addToVirtualClusters(portVc(cluster, "vc-relocate", relocateFromPort));
         try (KroxyliciousTester tester = KroxyliciousTesters.newBuilder(testerBuilder).createDefaultKroxyliciousTester()) {
@@ -477,7 +474,7 @@ class HotReloadIT extends BaseIT {
             assertProduceConsumeRoundTrip(tester, "vc-relocate", topicBefore, "given-old-port");
 
             // When
-            var afterConfig = portConfig(portVc(cluster, "vc-relocate", relocateToPort));
+            var afterConfig = portConfig(portVc(cluster, "vc-relocate", PORT_RELOCATE_TARGET_BOOTSTRAP));
             assertThat(tester.reconfigure(afterConfig))
                     .succeedsWithin(RECONFIGURE_TIMEOUT)
                     .satisfies(rr -> assertThat(rr.hasErrors())
@@ -654,9 +651,10 @@ class HotReloadIT extends BaseIT {
     }
 
     /**
-     * Polls until the port is bindable without SO_REUSEADDR. Using setReuseAddress(false)
-     * ensures we wait until the OS has fully released the socket, not just until TIME_WAIT
-     * allows a reuse-flagged bind to succeed.
+     * Polls until the port is bindable with SO_REUSEADDR, matching the semantics of the
+     * proxy's own {@code ServerBootstrap} ({@code SO_REUSEADDR=true}). This lets the probe
+     * succeed as soon as Netty closes the listening channel, without waiting for any
+     * TIME_WAIT connections to drain.
      */
     private static void assertPortIsBindable(int port) {
         await("port " + port + " to be bindable")
@@ -664,7 +662,7 @@ class HotReloadIT extends BaseIT {
                 .pollInterval(Duration.ofMillis(100))
                 .untilAsserted(() -> assertThatCode(() -> {
                     try (var s = new ServerSocket()) {
-                        s.setReuseAddress(false);
+                        s.setReuseAddress(true);
                         s.bind(new InetSocketAddress((InetAddress) null, port));
                     }
                 }).doesNotThrowAnyException());

